scio-elasticsearch/es8/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,755 @@
+        this.numOfShard = numOfShard;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        // assign this element to a random shard
+        final long shard = ThreadLocalRandom.current().nextLong(numOfShard);
+        c.output(KV.of(shard, c.element()));
+      }
+    }
+
+    private static class ElasticsearchWriter<T> extends DoFn<T, Void> {
+
+      private List<BulkOperation> chunk;
+      private long currentSize;
+      private long currentBytes;
+
+      private final ClientSupplier clientSupplier;
+      private final SerializableFunction<T, Iterable<BulkOperation>> toBulkOperations;
+      private final ThrowingConsumer<BulkExecutionException> error;
+      private final int maxBulkRequestSize;
+      private final long maxBulkRequestBytes;
+      private final int maxRetries;
+      private final Duration retryPause;
+
+      private ProcessFunction<BulkRequest, BulkResponse> requestFn;
+      private ProcessFunction<BulkRequest, BulkResponse> retryFn;
+
+      public ElasticsearchWriter(
+          HttpHost[] nodes,
+          int maxBulkRequestSize,
+          long maxBulkRequestBytes,
+          SerializableFunction<T, Iterable<BulkOperation>> toBulkOperations,
+          ThrowingConsumer<BulkExecutionException> error,
+          int maxRetries,
+          Duration retryPause,
+          UsernamePasswordCredentials credentials) {
+        this.maxBulkRequestSize = maxBulkRequestSize;
+        this.maxBulkRequestBytes = maxBulkRequestBytes;
+        this.clientSupplier = new ClientSupplier(nodes, credentials);
+        this.toBulkOperations = toBulkOperations;
+        this.error = error;
+        this.maxRetries = maxRetries;
+        this.retryPause = retryPause;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        checkArgument(this.clientSupplier.get().ping().value(),
+            "Elasticsearch client not reachable");
+
+        final FluentBackoff backoffConfig =
+            FluentBackoff.DEFAULT
+                .withMaxRetries(this.maxRetries)
+                .withInitialBackoff(this.retryPause);
+        this.requestFn = request(clientSupplier, error);
+        this.retryFn = retry(requestFn, backoffConfig);
+      }
+
+      @Teardown
+      public void teardown() throws Exception {
+        this.clientSupplier.close();
+      }
+
+      @StartBundle
+      public void startBundle(StartBundleContext c) {
+        chunk = new ArrayList<>();
+        currentSize = 0;
+        currentBytes = 0;
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        flush();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        final T t = c.element();
+        final Iterable<BulkOperation> operations = toBulkOperations.apply(t);
+        for (BulkOperation operation : operations) {
+          // long bytes = documentSize(operation);
```

Comment:
Since the operations are not serialized yet, it is hard to know the size at this point

Reply:
removing